Fix Protected Directory Detection and Admin Notice

CodeSmith32 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

1. make unprotected directory detection smarter: not every /woocommerce_uploads/ page with content is a directory listing
2. make notice actually close when clicking "Dismiss"

Closes #33436.

### How to test the changes in this Pull Request:

1. Modify `wp-content/uploads/woocommerce-uploads/index.html` (create it if it does not exist) and add some text  to this file (anything at all will do).
2. Reset your admin notices by clearing transients and calling `WC_Admin_Notices::reset_admin_notices()`. Additionally, a user meta field (`dismissed_uploads_directory_is_unprotected_notice`) may need to be deleted. To do this via WP CLI:
     - `wp transient delete _woocommerce_upload_directory_status`
     - `wp eval "WC_Admin_Notices::reset_admin_notices();"`
     - `wp user meta delete $(wp user get admin --field=ID) dismissed_uploads_directory_is_unprotected_notice` ← *change `admin` to the username of your test user.*
3. Now visit an admin screen such as **WooCommerce ▸ Settings ▸ General** and you should see an admin notice warning that your uploads directory is not protected:

![unprotected-downloads-dir](https://user-images.githubusercontent.com/3594411/204891850-2e8c698f-fe7f-4f2d-8faa-e22d3d282ecd.png)

4. Dismiss the notice ... give it a few seconds so that the page reloads and the notice should no longer be present.
5. Navigate to another admin screen such as **WooCommerce ▸ Settings ▸ Products**, then back to **WooCommerce ▸ Settings ▸ General**. The unprotected directory notice should remain hidden.
6. Once you complete your testing, you should restore `wp-content/uploads/woocommerce-uploads/index.html` to its earlier state (normally, this file is present but is completely empty).

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

CodeSmith32 committed:
Fix Protected Directory Admin Notice

- make notice actually close when clicking "Dismiss"
- make unprotected directory detection smarter: not every /woocommerce_uploads/ page with content is a directory listing

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-notices.php:
```diff
@@ -646,9 +646,16 @@ protected static function is_uploads_directory_protected() {
+		$is_protected = (
+			200 !== $response_code
+			|| empty( $response_content )
+			|| 1 !== preg_match( '/<title[^>]*>\\s*index\\s+of\\s+[\\s\\S]*<table/i', $response_content )
```
My only concern here is that, previously, we may erroneously have flagged that the uploads directory is unprotected when actually it **is** protected. That's not *great,* but at least it prompts the user to go take a look.

With this change, we avoid that issue in some % of cases, but we also risk **not** flagging that there is a problem when a directory listing page not meeting this regex is returned. That doesn't feel like the right trade-off, especially since (at least, following the adoption of the code in the second part of this PR) the admin notice can be dismissed.

CodeSmith32 replied:
I mainly added this because in my case I was additionally having this issue, which made it doubly obnoxious. You can refer to my comment on issue #26600 for the reason. But I personally think that issue should either *not* be closed, or should be closed with a 'wontfix' since even the OP references a site for which only something like this code would fix.

I agree that ideally the regex should cover every case. Therefore, I did do research to see if there were other types of index pages from different web engines (e.g., NGINX), and couldn't find any. I also researched to see if a translation or customization would throw off the regex, [but the Apache index page text appears to be hardcoded.](https://stackoverflow.com/questions/27018323/can-i-change-the-language-of-mod-autoindex-apache-directory-listing) So, I left it as is.

But if access to the `woocommerce_uploads` is such a concern, instead of checking if it's accessible or not, why doesn't the plugin just add a blank `index.php` to the folder when the plugin creates it? Wouldn't that be easier?

barryhughes replied:
> Since even the OP references a site for which only something like this code would fix

I'm not sure about that, though I may be misunderstanding or something may have changed since that was first posted. They reference [this URL](https://gallmet.hu/wp-content/uploads/woocommerce_uploads) and at time of writing it returns a 403 status—so I'd be surprised if that triggered the warning.

> I did do research to see if there were other types of index pages from different web engines.

Yep, and I appreciate the thorough research you did: I imagine the regex you wrote will cover *most* cases—but it still leaves others unaddressed (Nginx with `autoindex_format json`, various Lighttpd configs, a major non-English language host that has custom compiled Apache2).

Don't misunderstand, I fully understand that is likely a minority of users, but it still feels safer to give a false negative ('something is wrong, go check it out and dismiss this message if all is actually ok') than a false positive ('everything is just fine!' ... but isn't).

> But if access to the woocommerce_uploads is such a concern, instead of checking if it's accessible or not, why doesn't the plugin just add a blank index.php to the folder when the plugin creates it?

It already creates, or tries to create, both an `index.html` file and an `.htaccess` file. An `index.php` file could also be generated, but would require careful consideration first (since that could set off various alarm bells).

Looking at your note in #26599: is `index.html` missing in your case, or, is Nginx configured not to use `.html` files in this role?

CodeSmith32 replied:
> They reference [this URL](https://gallmet.hu/wp-content/uploads/woocommerce_uploads) and at time of writing it returns a 403 status—so I'd be surprised if that triggered the warning.

Hmm, yep, you're right! I missed that the status was 403, which would prevent it from being detected.

> Don't misunderstand, I fully understand that is likely a minority of users, but it still feels safer to give a false negative

And yes, I also agree! I was just under the impression that there were a lot more false positives, but...

> It already creates, or tries to create, both an index.html file and an .htaccess file.

Since this is true (I do see them on my server; should've checked first..), then I agree my case is specifically rare. I was under the impression the notice was to encourage the use of `Options -Indexes` without doing something to help prevent the issue. This clearly isn't the case.

> Looking at your note in https://github.com/woocommerce/woocommerce/issues/26599: is index.html missing in your case, or, is Nginx configured not to use .html files in this role?

The specific issue in my case is that I've modified my local system's 'hosts' file (a place for adding custom DNS overrides). Therefore, I'm able to visit the site, even though the domain points elsewhere. But because it points elsewhere, when WooCommerce makes the lookup, it uses the *actual* DNS and gets a different page than I do, which made it really confusing at first.

But if WC does create an index and even .htaccess block, what really would surprise me is finding a true positive case. But if indeed it is possible, then really feel free to leave the second half out. Being able to dismiss the message is certainly an acceptable compromise.

Thanks @barryhughes!

barryhughes replied:
> But if WC does create an index and even .htaccess block, what really would surprise me is finding a true positive case.

Yes, though, there are just so many installs out there and such a wide variety of environments with differing file permission issues and web server configs that the possibility is always there.

> Being able to dismiss the message is certainly an acceptable compromise.

Agree—in that case, would you be happy to revert/unwind this change (touching the `base_tables_missing_notice()` method)? I think we can then go ahead and merge the fix for notice dismissal.

barryhughes committed:
Unwind the change made to method `is_uploads_directory_protected()`.

Discussion here https://github.com/woocommerce/woocommerce/pull/33544/files#r910253861 relates. In essence, we felt it better to have more warnings prompting the user to evaluate if the directory is protected, than fewer warnings with an increased change of not flagging actual problems.

barryhughes committed:
Add changelog.

barryhughes commented:
- Updated per discussion.
- Testing instructions tweaked since the change is now smaller in scope.
- Waiting on checks passing before merge.
